doc: translation for docs/tutorials/timeseriees-forecasting-by-dsp.md

jxs1211 opened the pull request:
Signed-off-by: shenxianjie <327411586@qq.com>

#### What type of PR is this?
translate docs/tutorials/timeseriees-forecasting-by-dsp.md

#### What this PR does / why we need it:
fix the issue #421 

#### Which issue(s) this PR fixes:

Fixes #421

zsnmwy commented on docs/tutorials/timeseriees-forecasting-by-dsp.md:
```diff
@@ -1,104 +1,100 @@
+A fast discrete Fourier transform (FFT) is performed on the monitored time series (set length $N$) to obtain a spectrogram of the signal, which is visualized as "shocks" at each discrete point $k$. The height of the shock is the "amplitude" of the corresponding periodic component of $k$, and the value of $k$ ranges from $(0,1,2, ... N-1)$.
 
-$k = 0$对应信号的「直流分量」，对于周期没有影响，因此忽略。
+The "DC component" of the corresponding signal has no effect on the period and is therefore ignored.
```
```suggestion
The "DC component" of the corresponding signal does not affect the period, so $k = 0$ is ignored.
```

zsnmwy replied:
Don't miss the keyword.

zsnmwy commented on docs/tutorials/timeseriees-forecasting-by-dsp.md:
```diff
@@ -1,104 +1,100 @@
+The corresponding period $$T = {N \over k} \bullet SampleInterval$$
 
-要观察一个信号是不是以$T$为周期，至少需要观察两倍的$T$的长度，因此通过长度为$N$的序列能够识别出的最长周期为$N/2$。所以可以忽略$k = 1$。
+To see if a signal has a period of $$T$$, one needs to observe at least twice the length of $$T$$, so the longest period that can be identified by a sequence of length $$N$$ is $$N/2$$. So $k = 1$ can be ignored.
```
```suggestion
To see if a signal has a period of $$T$$, one needs to observe at least twice the length of $$T$$, so the longest period can be identified by a sequence of length $$N$$ is $$N/2$$. So $k = 1$ can be ignored.
```

zsnmwy commented on docs/tutorials/timeseriees-forecasting-by-dsp.md:
```diff
@@ -1,104 +1,100 @@
 
-至此，$k$的取值范围为$(2, 3, ... , N/2)$，对应的周期为$N/2, N/3, ...$，这也就是FFT能够提供的周期信息的「分辨率」。如果一个信号的周期没有落到$N/k$上，它会散布到整个频域，导致「频率泄漏」。
-好在在实际生产环境中，我们通常遇到的应用（尤其是在线业务），如果有规律，都是以「天」为周期的，某些业务可能会有所谓的「周末」效应，即周末和工作日不太一样，如果扩大到「周」的粒度去观察，它们同样具有良好的周期性。
+Up to this point, $k$ takes values in the range $(2, 3, ... , N/2)$, corresponding to the periods $N/2, N/3, ... $, which is the "resolution" of the period information that the FFT can provide. If the period of a signal does not fall on $N/k$, it will scatter to the whole frequency domain, leading to "frequency leakage". The good thing is that in a real production environment, the applications we ...
```
```suggestion
Up to this point, $k$ takes values in the range $(2, 3, ..., N/2)$, corresponding to the periods $N/2, N/3, ... $, which is the "resolution" of the period information that the FFT can provide. If the period of a signal does not fall on $N/k$, it will scatter to the whole frequency domain, leading to "frequency leakage". The good thing is that in a real production environment, the applications we usually encounter (especially online businesses), if there is regularity, are on a 'day' cycle, and some businesses may have the so-called 'weekend' effect, i.e., weekends are not quite the same as weekdays, and they are equally well periodic if expanded to the 'week' granularity of observation.
```

zsnmwy commented on docs/tutorials/timeseriees-forecasting-by-dsp.md:
```diff
@@ -1,104 +1,100 @@
+We find that the 'amplitude' at a few points is significantly higher than others, and these points can be used as our 'candidate cycles' for further verification.
 
-上面是我们通过直觉判断的，Crane是如何挑选「候选周期」的呢？
+The above is our intuitive judgment, how does Crane select the "candidate cycles"?
```
```suggestion
The above is our intuitive judgment. How does Crane select the "candidate cycles"?
```

zsnmwy commented on docs/tutorials/timeseriees-forecasting-by-dsp.md:
```diff
@@ -1,104 +1,100 @@
 ![](../images/algorithm/dsp/acf.png)
 
-Crane会依次验证每一个候选周期对应的自相关系数是否位于「山顶」上；并且选择对应「最高峰」的那个候选周期为整个时间序列的主周期（基波周期），并以此为基础进行预测。
+Crane will verify whether the autocorrelation coefficient of each candidate period is on the "peak" in turn; and select the candidate period corresponding to the "highest peak" as the main period (fundamental period) of the whole time series, and use it as the basis for prediction.
```
```suggestion
Crane will verify whether the autocorrelation coefficient of each candidate period is on the "peak" in turn, select the candidate period corresponding to the "highest peak" as the main period (fundamental period) of the whole time series, and use it as the basis for prediction.
```

zsnmwy commented on docs/tutorials/timeseriees-forecasting-by-dsp.md:
```diff
@@ -1,104 +1,100 @@
+ **fft**
 
-对原始时间序列做FFT得到频谱序列，去除「高频噪声」后，再做IFFT（逆快速傅里叶变换），将得到的时间序列作为下一个周期的预测结果。
+Do FFT on the original time series to get the spectral series, remove the "high frequency noise", then do IFFT (Inverse Fast Fourier Transform), and use the obtained time series as the prediction result for the next cycle.
```
```suggestion
Do FFT on the original time series to get the spectral series, remove the "high-frequency noise", then do IFFT (Inverse Fast Fourier Transform), and use the obtained time series as the prediction result for the next cycle.
```

zsnmwy commented on docs/tutorials/timeseriees-forecasting-by-dsp.md:
```diff
@@ -111,71 +107,69 @@ spec:
-#              - marginFraction: "0.1"
+          sampleInterval: "60s" # The sampling interval for monitoring data is 1 minute
+          historyLength: "15d" # Pull the last 15 days of monitoring metrics as the basis for prediction
+          estimators: # Specify the prediction method, including 'maxValue' and 'fft', each category can specify multiple estimators with different parameters, and crane will select the one with the best fit to generate the prediction results. If not specified, 'fft' will be used by default.
+# maxValue:
+# - marginFraction: "0.1"
             fft:
               - marginFraction: "0.2"
                 lowAmplitudeThreshold: "1.0"
                 highFrequencyThreshold: "0.05"
                 minNumOfSpectrumItems: 10
                 maxNumOfSpectrumItems: 20
```
indent error
<img width="810" alt="image" src="https://user-images.githubusercontent.com/35299017/182680454-26f743e2-a4cc-4fb7-8651-709e0906583a.png">
<img width="724" alt="image" src="https://user-images.githubusercontent.com/35299017/182680508-bba83fc7-ef2f-40ff-86f7-90f173ba0513.png">

jxs1211 replied:
fixed

zsnmwy commented on docs/tutorials/timeseriees-forecasting-by-dsp.md:
```diff
@@ -111,71 +107,69 @@ spec:
+`lowAmplitudeThreshold`: The lower limit of the amplitude of the spectrum, all frequency components with amplitudes below this lower limit will be filtered out.
 
-`highFrequencyThreshold`: 频率上限，所有频率高于这个上限的频率分量将被滤除。单位Hz，例如如果想忽略长度小于1小时的周期分量，设置`highFrequencyThreshold = 1/3600`。
+`highFrequencyThreshold`: The upper frequency limit, all frequencies above this limit will be filtered out. Unit Hz, for example, if you want to ignore the length of the period component less than 1 hour, set `highFrequencyThreshold = 1/3600`.
```
```suggestion
`highFrequencyThreshold`: The upper-frequency limit, all frequencies above this limit will be filtered out. Unit Hz, for example, if you want to ignore the length of the period component less than 1 hour, set `highFrequencyThreshold = 1/3600`.
```

jxs1211 replied:
fixed

zsnmwy commented on docs/tutorials/timeseriees-forecasting-by-dsp.md:
```diff
@@ -111,71 +107,69 @@ spec:
+`maxNumOfSpectrumItems`: the maximum number of frequency components to keep.
 
-简单来说，保留频率分量的数量越少、频率上限越低、频谱幅度下限越高，预测出来的曲线越光滑，但会丢失一些细节；反之，曲线毛刺越多，保留更多细节。
+In simple terms, the lower the number of reserved frequency components, the lower the upper frequency limit, and the higher the lower limit of spectral amplitude, the smoother the predicted curve will be, but some details will be lost; conversely, the more burrs the curve will have, the more details will be preserved.
```
```suggestion
In simple terms, the lower the number of reserved frequency components, the lower the upper-frequency limit, and the higher the lower limit of spectral amplitude, the smoother the predicted curve will be, but some details will be lost; conversely, the more burrs the curve will have, the more details will be preserved.
```

jxs1211 replied:
fixed

zsnmwy commented on docs/tutorials/timeseriees-forecasting-by-dsp.md:
```diff
@@ -111,71 +107,69 @@ spec:
+In simple terms, the lower the number of reserved frequency components, the lower the upper frequency limit, and the higher the lower limit of spectral amplitude, the smoother the predicted curve will be, but some details will be lost; conversely, the more burrs the curve will have, the more details will be preserved.
 
-下面是对同一时段预测的两条曲线，蓝色、绿色的`highFrequencyThreshold`分别为$0.01$和$0.001$，蓝色曲线过滤掉了更多的高频分量，因此更为平滑。
+Below are two curves predicted for the same time period, the blue and green `highFrequencyThreshold` is $0.01$ and $0.001$ respectively, the blue curve filters out more high frequency components, so it is smoother.
```
```suggestion
Below are two curves predicted for the same time period, the blue and green `highFrequencyThreshold` is $0.01$ and $0.001$ respectively, and the blue curve filters out more high-frequency components, so it is smoother.
```

jxs1211 replied:
fixed

zsnmwy commented on docs/tutorials/timeseriees-forecasting-by-dsp.zh.md:
```diff
@@ -0,0 +1,181 @@
+
+![](../images/algorithm/dsp/missing_data_fill.png)
+#### 去除异常点
+监控数据中偶尔会出现一些极端的异常数据点，导致这些异常点（outliers）的原因有很多，例如：
+1. 监控系统用0值填充缺失的采样点；
+2. 被监控组件由于自身的bug上报了错误的指标数据；
+3. 应用启动时会消耗远超正常运行时的资源
```
```suggestion
监控数据中偶尔会出现一些极端的异常数据点，导致这些异常点（outliers）的原因有很多，例如：

1. 监控系统用0值填充缺失的采样点；
2. 被监控组件由于自身的bug上报了错误的指标数据；
3. 应用启动时会消耗远超正常运行时的资源
```
Add a new line to make the list style work.

jxs1211 replied:
fixed

zsnmwy commented:
Thanks for your pull request. :)

jxs1211 commented on docs/tutorials/timeseriees-forecasting-by-dsp.md:
```diff
@@ -1,104 +1,100 @@
+It is common for monitoring data to be missing at certain points in time. Crane will fill in the missing sampling points based on the before and after data. The approach is as follows.
 
-假设第$m$个与第$n$个采样点之间采样数据缺失（$m+1 < n$）,设在$m$和$n$点的采样值分别为$v_m$和$v_n$，令$$\Delta = {v_n-v_m \over n-m}$$，则$m$和$n$之间的填充数据依次为$v_m+\Delta , v_m+2\Delta , ...$
+Assume that the data sampled between the $m$th and $n$th sample points are missing ($m+1 < n$), and the sample values at the $m$ and $n$ points are $v_m$ and $v_n$ respectively, so that $$\Delta = {v_n-v_m \over n-m}$$, then the filled data between $m$ and $n$ are $$v_m+\Delta , v_m+2\ Delta , ... $
```
fixed

jxs1211 commented on docs/tutorials/timeseriees-forecasting-by-dsp.md:
```diff
@@ -1,104 +1,100 @@
+Crane uses the discrete Fourier transform and autocorrelation function commonly used in the field of Digital Signal Processing (DSP) to identify and predict periodic time series.
 
-本文将介绍DSP算法的实现流程和参数设置，以便帮助大家了解算法背后的原理，并将它应用到实际场景中。 （相关代码位于`pkg/prediction/dsp`目录下）
+This article will introduce the implementation process and parameter settings of the DSP algorithm in order to help you understand the principles behind the algorithm and apply it to practical scenarios. (The related code is located in the pkg/prediction/dsp directory)
```
fixed

jxs1211 commented on docs/tutorials/timeseriees-forecasting-by-dsp.md:
```diff
@@ -1,104 +1,100 @@
-#### 填充缺失数据
-监控数据在某些时间点上缺失是很常见的现象，Crane会根据前后的数据对缺失的采样点进行填充。做法如下：
+#### Fill in missing data
+It is common for monitoring data to be missing at certain points in time. Crane will fill in the missing sampling points based on the before and after data. The approach is as follows.
```
fixed

jxs1211 commented on docs/tutorials/timeseriees-forecasting-by-dsp.md:
```diff
@@ -1,104 +1,100 @@
+Occasionally, there are extreme outlier data points in the monitoring data. There are many reasons for these outliers, such as:
 
-这些极端的异常点对于信号的周期判断会造成干扰，需要进行去除。做法如下：
+1. Monitoring systems filling missing sample points with zero values.
```
fixed

jxs1211 commented on docs/tutorials/timeseriees-forecasting-by-dsp.md:
```diff
@@ -1,104 +1,100 @@
+3. Application startup consumes far more resources than normal runtime
 
-选取实际序列中所有采样点的$P99.9$和$P0.1$，分别作为上、下限阈值，如果某个采样值低于下限或者高于上限，将采样点的值设置为前一个采样值。
+These extreme anomalies can be disruptive to the signal cycle determination and need to be removed. The approach is as follows:
```
fixed

jxs1211 commented:
> Thanks for your pull request. :)

@zsnmwy
I have fixed all the comments.
There is a problem here, I firstly commit the pr on a computer, but now I notice I can't get the content of my first commit when I am using another computer after pulling my forked branch.
I found out the commit's SHA and attach it to it, then the first commit come back, 
`
[going@dev crane]$ git branch
* (HEAD detached from f33e209)
  main
`
so I fixed and update the tag, Now I have no idea how to push the pr in the right way, I try to push it as below:
`[going@dev crane]$ git remote -v
origin  git@github.com:jxs1211/crane.git (fetch)
origin  git@github.com:jxs1211/crane.git (push)
upstream        git@github.com:gocrane/crane.git (fetch)
upstream        no_push (push)
[going@dev crane]$ git push origin f33e209:main
`
but it seems don't trigger the ci. Could you do me a favor to fix the situation?

qmhu commented:
> > Thanks for your pull request. :)
> 
> @zsnmwy I have fixed all the comments. There is a problem here, I firstly commit the pr on a computer, but now I notice I can't get the content of my first commit when I am using another computer after pulling the forked repo. I found out the commit's SHA and attach to it, then the first commit come back,
> 
> ```
> [going@dev crane]$ git branch
> * (HEAD detached from f33e209)
>   main
> ```
> 
> so I fixed and update the tag, Now I have no idea how to push the pr in the right way, I try to push it as below:
> 
> ```
> origin  git@github.com:jxs1211/crane.git (fetch)
> origin  git@github.com:jxs1211/crane.git (push)
> upstream        git@github.com:gocrane/crane.git (fetch)
> upstream        no_push (push)
> [going@dev crane]$ git push origin f33e209:main
> ```
> 
> but it seems don't trigger the ci. Could you do me a favor to resolve the situation?

如果弄混了，可以这样：
重新clone一个文件夹，checkout到你的branch，本地准备好你的提交内容，git commit，git push -f origin {your-branch}

jxs1211 commented:
> > > Thanks for your pull request. :)
> > 
> > 
> > @zsnmwy I have fixed all the comments. There is a problem here, I firstly commit the pr on a computer, but now I notice I can't get the content of my first commit when I am using another computer after pulling the forked repo. I found out the commit's SHA and attach to it, then the first commit come back,
> > ```
> > [going@dev crane]$ git branch
> > * (HEAD detached from f33e209)
> >   main
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > so I fixed and update the tag, Now I have no idea how to push the pr in the right way, I try to push it as below:
> > ```
> > origin  git@github.com:jxs1211/crane.git (fetch)
> > origin  git@github.com:jxs1211/crane.git (push)
> > upstream        git@github.com:gocrane/crane.git (fetch)
> > upstream        no_push (push)
> > [going@dev crane]$ git push origin f33e209:main
> > ```
> > but it seems don't trigger the ci. Could you do me a favor to resolve the situation?
> 
> 如果弄混了，可以这样： 重新clone一个文件夹，checkout到你的branch，本地准备好你的提交内容，git commit，git push -f origin {your-branch}

@zsnmwy @qmhu I've opened another pr# for this issue, please close this pr.

zsnmwy commented:
> > > > Thanks for your pull request. :)
> > > 
> > > 
> > > @zsnmwy I have fixed all the comments. There is a problem here, I firstly commit the pr on a computer, but now I notice I can't get the content of my first commit when I am using another computer after pulling the forked repo. I found out the commit's SHA and attach to it, then the first commit come back,
> > > ```
> > > [going@dev crane]$ git branch
> > > * (HEAD detached from f33e209)
> > >   main
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > so I fixed and update the tag, Now I have no idea how to push the pr in the right way, I try to push it as below:
> > > ```
> > > origin  git@github.com:jxs1211/crane.git (fetch)
> > > origin  git@github.com:jxs1211/crane.git (push)
> > > upstream        git@github.com:gocrane/crane.git (fetch)
> > > upstream        no_push (push)
> > > [going@dev crane]$ git push origin f33e209:main
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > but it seems don't trigger the ci. Could you do me a favor to resolve the situation?
> > 
> > 
> > 如果弄混了，可以这样： 重新clone一个文件夹，checkout到你的branch，本地准备好你的提交内容，git commit，git push -f origin {your-branch}
> 
> @zsnmwy @qmhu I've opened another pr#452 for this issue, please close this pr.

Not the same content with #451 and #452. Is right?

zsnmwy commented:
> > Thanks for your pull request. :)
> 
> @zsnmwy I have fixed all the comments. There is a problem here, I firstly commit the pr on a computer, but now I notice I can't get the content of my first commit when I am using another computer after pulling the forked repo. I found out the commit's SHA and attach to it, then the first commit come back,
> 
> ```
> [going@dev crane]$ git branch
> * (HEAD detached from f33e209)
>   main
> ```
> 
> so I fixed and update the tag, Now I have no idea how to push the pr in the right way, I try to push it as below:
> 
> ```
> origin  git@github.com:jxs1211/crane.git (fetch)
> origin  git@github.com:jxs1211/crane.git (push)
> upstream        git@github.com:gocrane/crane.git (fetch)
> upstream        no_push (push)
> [going@dev crane]$ git push origin f33e209:main
> ```
> 
> but it seems don't trigger the ci. Could you do me a favor to resolve the situation?

The commit f33e209 is already pushed to jxs1211/crane:main.
Nothing will be changed if you try to push the same commit twice because the commit is already in GitHub.
Push the same commit can't trigger the CI.

1. You can use the `git stash` to save your work - docs fix.
2. Use Github official CLI to get the pr content. `gh pr checkout 451`. The CLI just wrapper git command.
3. Use `git stash pop` to recover your work.
4. Follow the after guide.

Ref: [git stash](https://git-scm.com/docs/git-stash)

---

If you want to make some content changes, I recommend the two methods.
- standard dev.
  ```bash
  git checkout jxs1211/doc-translation-timeseriees-forecasting-by-dsp
  echo "A" >> test.md
  git add .
  git commit -m "Add A" -s
  git push
  ```
- Changing the Last Commit with amend
  ```bash
  git checkout jxs1211/doc-translation-timeseriees-forecasting-by-dsp
  # Edit hello.py and main.py
  git add hello.py
  git commit # Here, maybe get the commit SHA - 123123. 
  git push # Now you can see the commit hash - 123123 on the Github page.
  # Realize you forgot to add the changes from main.py or want to fix some problems.
  git add main.py 
  git commit --amend --no-edit # Here, maybe get the commit SHA - 456456
  # Notice the git history is changed. You must use the `--force` params when using git push. Replace the remote git history with the local git history. You must ensure the remote repo history no one use or changes.
  git push --force # Now you can see the new commit hash - 456456 on the Github page. The remote git history will miss the hash 123123.
  ```
  **You must ensure the remote repo history no one use or changes.**
  **You must ensure the remote repo history no one use or changes.**
  **You must ensure the remote repo history no one use or changes.**

Ref: [git amend](https://www.atlassian.com/git/tutorials/rewriting-history)
